File: client/src/game/engine/missionCompletionCalculator.ts
// src/game/engine/missionCompletionCalculator.ts

import Side from "@/game/Side.ts";
import Scenario from "@/game/Scenario";
import PatrolMission from "@/game/mission/PatrolMission";
import StrikeMission from "@/game/mission/StrikeMission";

/**
 * Checks if the side has completed all assigned missions
 * @param Scenario current scenario object
 * @param sideId id of concerned side
 * @returns boolean
*/
export function areAllMissionsCompleteFromScenario(scenario: Scenario, sideId: string): boolean {
  const side = scenario.getSide(sideId);
  if (side && side.missionsCompleted === side.missionsAssigned) {
    return true;
  }

  return false;
}

/**
 * Checks if the side has completed all assigned missions
 * @param Side object
 * @returns boolean all assigned missions are complete 
 */
export function areAllMissionsCompleteFromSide(side: Side): boolean {
  
  if (side 
    && side.missionsCompleted > 0
    && side.missionsCompleted === side.missionsAssigned) {
    return true;
  }

  return false;
}

/**
 * Increments the success and completion counters for a side
 * after its strike mission succeeds (target destroyed).
 * @param scenario The current game scenario.
 * @param mission The successful strike mission.
 */
export function incrementStrikeMissionSuccess(scenario: Scenario, mission: StrikeMission) {
  const side = scenario.getSide(mission.sideId);
  if (side) {
    side.missionsCompleted = (side.missionsCompleted || 0) + 1;
    side.missionsSucceeded = (side.missionsSucceeded || 0) + 1;
    // console.log(`${side.name} completed a successful strike mission. Total Succeeded: ${side.missionsSucceeded}`);
}
}

/**
 * Increments the failure and completion counters for a side
 * after its strike mission fails (e.g., all attackers lost).
 * @param scenario The current game scenario.
 * @param mission The failed strike mission.
 */
export function incrementStrikeMissionFailure(scenario: Scenario, mission: StrikeMission) {
  const side = scenario.getSide(mission.sideId);
  if (side) {
    side.missionsCompleted = (side.missionsCompleted || 0) + 1;
    side.missionsFailed = (side.missionsFailed || 0) + 1;
    // console.log(`${side.name} failed a strike mission. Total Failed: ${side.missionsFailed}`);
  }
}


/**
 * Increments the success and completion counters for a side
 * after its patrol mission succeeds.
 * @param scenario The current game scenario.
 * @param mission The successful patrol mission.
 */
export function incrementPatrolMissionSuccess(scenario: Scenario, mission: PatrolMission) {
  const side = scenario.getSide(mission.sideId);
  if (side) {
    side.missionsCompleted = (side.missionsCompleted || 0) + 1; // This line was added
    side.missionsSucceeded = (side.missionsSucceeded || 0) + 1;
    // console.log(`${side.name} completed a successful patrol mission. Total Succeeded: ${side.missionsSucceeded}`);
  }
}

/**
 * Increments the failure and completion counters for a side
 * after its patrol mission fails (all assigned units lost).
 * @param scenario The current game scenario.
 * @param mission The failed patrol mission.
 */
export function incrementPatrolMissionFailure(scenario: Scenario, mission: PatrolMission) {
  const side = scenario.getSide(mission.sideId);
  if (side) {
    side.missionsCompleted = (side.missionsCompleted || 0) + 1;
    side.missionsFailed = (side.missionsFailed || 0) + 1;
    // console.log(`${side.name} failed a patrol mission. Total Failed: ${side.missionsFailed}`);
  }
}

/**
 * Calculates the mission success rate for a Side. 
 * The calculation is as follows:
 *     missionSuccessRate = missionSuccessCount / totalCompletedMissionCount (arbitrary)
 * This function uses live Scenario instance.
 * @param scenario The final scenario state as a plain object.
 * @param sideId The ID of the side to calculate the rate for.
 * @returns The mission success rate as a percentage (0-100).
 */
export function calculateSideMissionSuccessRate(side: Side): number {

  if (!side || !side.missionsCompleted || side.missionsCompleted === 0) {
    return 0.0;
  }
    
  return side.missionsSucceeded / side.missionsCompleted;
}

/**
 * Calculates the mission success rate for a Side. 
 * The calculation is as follows:
 *     missionSuccessRate = missionSuccessCount / totalCompletedMissionCount (arbitrary)
 * This function is designed to work with a plain JSON object, not a Scenario class instance.
 * @param scenario The final scenario state as a plain object.
 * @param sideId The ID of the side to calculate the rate for.
 * @returns The mission success rate as a percentage (0-100).
 */
export function calculateMissionSuccessRateFromObject(scenario: any, sideId: string): number {
  const side = scenario.sides.find((s: any) => s.id === sideId);

  if (!side || !side.missionsCompleted || side.missionsCompleted === 0) {
    return 0.0;
  }

  return side.missionsSucceeded / side.missionsCompleted;
}

File: gym/blade/engine/missionCompletionCalculator.py
from typing import Dict, Any, List, Optional

from sympy import false, true
from blade.Side import Side
from blade.Scenario import Scenario
from blade.mission.PatrolMission import PatrolMission
from blade.mission.StrikeMission import StrikeMission

def are_all_missions_complete(side: Side):
    """
    Checks if the side has completed all assigned missions
    Takes in Side
    Returns boolean
    """
    if side and side.missions_completed == side.missions_assigned:
        return true
    
    return false
        

# --- Mission Success/Failure Counters ---

def increment_strike_mission_success(scenario: Scenario, mission: StrikeMission):
    """
    Increments the success and completion counters for a side
    after its strike mission succeeds.
    """
    side = scenario.get_side(mission.side_id)
    if side:
        side.missions_completed = getattr(side, 'missions_completed', 0) + 1
        side.missions_succeeded = getattr(side, 'missions_succeeded', 0) + 1
        print(f"{side.name} completed a successful strike mission. Total Succeeded: {side.missions_succeeded}")

def increment_strike_mission_failure(scenario: Scenario, mission: StrikeMission):
    """
    Increments the failure and completion counters for a side
    after its strike mission fails.
    """
    side = scenario.get_side(mission.side_id)
    if side:
        side.missions_completed = getattr(side, 'missions_completed', 0) + 1
        side.missions_failed = getattr(side, 'missions_failed', 0) + 1
        print(f"{side.name} failed a strike mission. Total Failed: {side.missions_failed}")

def increment_patrol_mission_success(scenario: Scenario, mission: PatrolMission):
    """
    Increments the success and completion counters for a side
    after its patrol mission succeeds.
    """
    side = scenario.get_side(mission.side_id)
    if side:
        side.missions_completed = getattr(side, 'missions_completed', 0) + 1
        side.missions_succeeded = getattr(side, 'missions_succeeded', 0) + 1
        print(f"{side.name} completed a successful patrol mission. Total Succeeded: {side.missions_succeeded}")

def increment_patrol_mission_failure(scenario: Scenario, mission: PatrolMission):
    """
    Increments the failure and completion counters for a side
    after its patrol mission fails.
    """
    side = scenario.get_side(mission.side_id)
    if side:
        side.missions_completed = getattr(side, 'missions_completed', 0) + 1
        side.missions_failed = getattr(side, 'missions_failed', 0) + 1
        print(f"{side.name} failed a patrol mission. Total Failed: {side.missions_failed}")

# --- Success Rate Calculators ---

def calculate_side_mission_success_rate(side: Side) -> float:
    """
    Calculates the mission success rate for a Side using live Scenario instances.
    Returns the rate as a float between 0.0 and 1.0.
    """
    if not side or getattr(side, 'missions_completed', 0) == 0:
        return 0.0
    
    missions_succeeded = getattr(side, 'missions_succeeded', 0)
    missions_completed = getattr(side, 'missions_completed', 0)
    
    return missions_succeeded / missions_completed

def calculate_mission_success_rate_from_object(scenario_obj: Dict[str, Any], side_id: str) -> float:
    """
    Calculates mission success rate from a plain dictionary representation of a scenario.
    Returns the rate as a float between 0.0 and 1.0.
    """
    side_data = next((s for s in scenario_obj.get("sides", []) if s.get("id") == side_id), None)

    if not side_data:
        return 0.0
    
    missions_completed = side_data.get("missionsCompleted", 0)
    missions_succeeded = side_data.get("missionsSucceeded", 0)

    if missions_completed == 0:
        return 0.0
        
    return missions_succeeded / missions_completed
